### PR TITLE
Issue/3083 modules v2 pypi package source

### LINKED ISSUE #3083: Modules V2 - pypi package source
- [x] Add the pypi package source, that can use pip to install modules 
- [x] integrate new install flow into compiler
- [x] #3087 moved `inmanta module install` behavior to `inmanta project install`. This currently only installs v1 modules but should be modified to install v2 as well. This bullet may be moved to #3051 if deemed appropriate.
- [x] make sure pypi provider can only produce v2 modules and other providers can only provide v1 modules.

### PR DESCRIPTION
**Pull request by sanderr**
# Description

Still in draft because it requires a lot more tests. All existing tests are succeeding at the moment of writing so the code itself should be pretty representative.

closes #3083

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature) => additional tickets are being created
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (#3200)

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/module.py [hunk 1]
```diff
@@ -226,14 +232,110 @@ def get_file_for_version(self, repo: str, tag: str, file: str) -> str:
...
     def path_for(self, name: str) -> Optional[str]:
-        # same class is used for search parh and remote repos, perhaps not optimal
+        if name.startswith(ModuleV2.PKG_NAME_PREFIX):
+            raise Exception("PythonRepo instances work with inmanta module names, not Python package names.")
+        package: str = f"inmanta_plugins.{name}"
+        init: Optional[str] = env.ProcessEnv.get_module_file(package)
+        if init is None:
+            return None
+        try:
+            return ModuleLike.get_first_directory_containing_file(os.path.dirname(init), ModuleV2.MODULE_FILE)
```
Comment (wouterdb) [inline src/inmanta/module.py:308]: this I don't understand

Comment (sanderr) [reply src/inmanta/module.py:308]: Do you mean the whole method logic? It checks if the `inmanta_plugins.the_module` package can be found in the active venv. If so, it goes up the directory tree starting from the `__init__.py` file until it finds a `ModuleV2.MODULE_FILE`, indicating the module root. It would be cleaner to use `return ModuleV2.get_module_dir(os.path.dirname(init))` here.

Comment (wouterdb) [reply src/inmanta/module.py:308]: The walking up is a bit weird here. Either the file is exactly where it is expected or the module is corrupt?

Comment (sanderr) [reply src/inmanta/module.py:308]: The thing is that it's different for editable vs non-editable installs, which is why I thought it easiest to determine the module root by its defining config file.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/env.py [hunk 2]
```diff
@@ -57,7 +66,43 @@ class ProcessEnv:
...
+                    "-m",
+                    "pip",
+                    "install",
+                    *(["--upgrade"] if upgrade else []),
+                    *(str(requirement) for requirement in requirements),
+                    *index_args,
+                ],
+                stderr=subprocess.PIPE,
+            )
+            process.check_returncode()
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/env.py:95]: Because I only wanted to capture stderr.

File: src/inmanta/module.py [hunk 3]
```diff
@@ -910,6 +1012,13 @@ def set(cls, project: "Project") -> None:
         plugins.PluginMeta.clear()
         loader.unload_inmanta_plugins()
 
+    def install_modules(self) -> None:
+        """
+        Installs all modules, both v1 and v2.
+        """
+        self.load_module_recursive(install=True)
+        self.load()
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/module.py:1020]: To pass `install=True`, though that could also be achieved by adding the parameter to `load` and `get_complete_ast` to pass it through if you prefer.

File: src/inmanta/module.py [hunk 4]
```diff
@@ -835,9 +933,13 @@ def __init__(self, path: str, autostd: bool = True, main_file: str = "main.cf",
         self.main_file = main_file
 
         self._metadata.modulepath = [os.path.abspath(os.path.join(path, x)) for x in self._metadata.modulepath]
-        self.resolver = CompositeModuleRepo([make_repo(x) for x in self.modulepath])
-        self.repolist = [x.url for x in self._metadata.repo]
-        self.externalResolver = CompositeModuleRepo([make_repo(x, root=path) for x in self.repolist])
+        self.module_source: ModuleV2Source = ModuleV2Source(
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/module.py:936]: This is something I went back and forth on. My latest reasoning was that this is now the main module source, v1 being the legacy one, so we needn't name it v2 explicitly. I don't feel this very strongly so I'm willing to change it if you think it's clearer.

File: src/inmanta/module.py (hunk 1 above)
Comment (sanderr) [reply src/inmanta/module.py]: I've updated this to be like Arnaud did it in #3164.

File: src/inmanta/env.py [hunk 5]
```diff
@@ -58,7 +67,53 @@ class ProcessEnv:
...
+            # TODO: other install methods patch pkg_resources.working_set, do we need to do so here as well?
```
Comment (wouterdb) [inline src/inmanta/env.py:106]: yes (and then some for the namespaces, I think you are around made a method for that)

File: src/inmanta/env.py [hunk 6]
```diff
@@ -69,17 +124,106 @@ def install_from_source(cls, paths: List[LocalPackagePath]) -> None:
             raise Exception("install_from_source requires at least one package to install")
         # make sure we only try to install from a local source
         explicit_paths: Iterator[LocalPackagePath] = (
-            LocalPackagePath(path=os.path.join(".", path.path), editable=path.editable) for path in paths
+            LocalPackagePath(path=os.path.join(".", path.path, ""), editable=path.editable) for path in paths
```
Comment (wouterdb) [inline src/inmanta/env.py:127]: why `,""`

File: src/inmanta/env.py [hunk 7]
```diff
@@ -69,17 +124,106 @@ def install_from_source(cls, paths: List[LocalPackagePath]) -> None:
...
             [
-                cls.env_path,
+                cls.python_path,
                 "-m",
                 "pip",
                 "install",
                 *chain.from_iterable(["-e", path.path] if path.editable else [path.path] for path in explicit_paths),
             ]
         )
+        # TODO: other install methods patch pkg_resources.working_set, do we need to do so here as well?
```
Comment (wouterdb) [inline src/inmanta/env.py:138]: yes

File: src/inmanta/env.py [hunk 8]
```diff
@@ -69,17 +124,106 @@ def install_from_source(cls, paths: List[LocalPackagePath]) -> None:
...
+
+        :param in_scope: A full pattern representing the package names that are considered in scope for the installed packages'
+            compatibility check. If constraints are supplied, those are always checked, regardless of this pattern.
+        :param constraints: In addition to checking for compatibility within the environment, also verify that the environment's
+            packages meet the given constraints. This is insterpreted as a list of constraints, not a list of requirements: if
+            one of the constraints' package is not installed the check succeeds.
+        """
+
+        def check_installed() -> List[str]:
+            process: subprocess.CompletedProcess = subprocess.run(
```
Comment (wouterdb) [inline src/inmanta/env.py:153]: This is called on every compile, it would be good to investigate if we can do this faster via, e.g. pkg_resources


(logically, all this info has to be in the loader already, so we shouldn't have to fork for it)

File: src/inmanta/env.py [hunk 9]
```diff
@@ -69,17 +124,106 @@ def install_from_source(cls, paths: List[LocalPackagePath]) -> None:
...
+            else:
+                process.check_returncode()
+            return []
+
+        def check_constraints() -> Iterator[Tuple[Requirement, version.Version]]:
+            if constraints is None:
+                return iter(())
+
+            installed_packages: Dict[str, version.Version] = {
+                name.lower(): version for name, version in get_installed_packages(cls.python_path).items()
```
Comment (wouterdb) [inline src/inmanta/env.py:177]: This is another fork I would very much try to avoid.

File: src/inmanta/env.py [hunk 10]
```diff
@@ -69,17 +124,106 @@ def install_from_source(cls, paths: List[LocalPackagePath]) -> None:
...
+        try:
+            spec = importlib.util.find_spec(module)
+        # inmanta.loader.PluginModuleLoader raises ImportError if module is not found
+        except (ImportError, ModuleNotFoundError):
+            spec = None
+        # TODO: test with v1 loader
+        return (spec.origin, spec.loader) if spec is not None else None
+
+    @classmethod
+    def init_namespace(cls, namespace: str) -> None:
```
Comment (wouterdb) [inline src/inmanta/env.py:211]: I think @arnaudsjs solved the problem of the namespace by adding it to the core package

File: src/inmanta/env.py [hunk 11]
```diff
@@ -101,7 +245,7 @@ def __init__(self, env_path: str) -> None:
         self._site_packages_dir: Optional[str] = None
 
     def get_package_installed_in_parent_env(self) -> Optional[Dict[str, str]]:
```
Comment (wouterdb) [inline src/inmanta/env.py:247]: is this dead code now?

File: src/inmanta/loader.py [hunk 12]
```diff
@@ -95,10 +95,11 @@ def get_siblings(self) -> Iterator["SourceInfo"]:
     @property
     def requires(self) -> List[str]:
         """List of python requirements associated with this source file"""
-        from inmanta.module import Project
+        from inmanta.module import Module, ModuleV1, Project
 
         if self._requires is None:
-            self._requires = Project.get().modules[self._get_module_name()].get_python_requirements_as_list()
+            module: Module = Project.get().modules[self._get_module_name()]
+            self._requires = module.get_python_requirements_as_list() if isinstance(module, ModuleV1) else []
```
Comment (wouterdb) [inline src/inmanta/loader.py:102]: would it not be cleaner to have this method on both module types, where a v2 always return []?

**Review by wouterdb — COMMENTED** [review]
I'm pushing results a few files at a time

File: src/inmanta/module.py [hunk 13]
```diff
@@ -1205,58 +1442,82 @@ def requires(self) -> "List[Requirement]":
...
+        for name, module in self.modules.items():
+            if module.GENERATION == ModuleGeneration.V1 and self.module_source.path_for(name) is not None:
+                LOGGER.warning(
+                    "Module %s was installed installed as v1 but later pulled in as v2 a dependency as v2."
+                    " Continuing would lead to inconsistent behavior. A rerun should resolve this issue.",
+                    name,
+                )
+                good = False
+
+        requirements: Dict[str, List[InmantaModuleRequirement]] = self.collect_requirements()
```
Comment (wouterdb) [inline src/inmanta/module.py:1491]: Would it not be simpler to 
- have every module know its own version 
- every module can return its module requirements/constraints
- evaluate the constraints against the set of loaded versions?

File: src/inmanta/module.py [hunk 14]
```diff
@@ -276,14 +320,155 @@ def get_file_for_version(self, repo: str, tag: str, file: str) -> str:
...
+    def clone(self, name: str, dest: str) -> bool:
+        raise NotImplementedError("Abstract method")
+
+    def install(self, project: "Project", module_spec: List[InmantaModuleRequirement]) -> Optional["ModuleV1"]:
+        module_name: str = self._get_module_name(module_spec)
+        path: Optional[str] = self.path_for(module_name)
+        if path is not None:
+            return self.from_path(project, path)
+        else:
+            return ModuleV1.install(project, module_name, module_spec, install_mode=project.install_mode)
```
Comment (wouterdb) [inline src/inmanta/module.py:466]: Why not move this method from ModuleV1 to here?

**Review by wouterdb — COMMENTED** [review]
next batch

File: src/inmanta/env.py (hunk 6 above)
Comment (sanderr) [reply src/inmanta/env.py:127]: To make sure a `/` is added at the end. Otherwise pip doesn't look for a local path iirc.

File: src/inmanta/env.py (hunk 10 above)
Comment (sanderr) [reply src/inmanta/env.py:211]: This was definitely causing me issues (actually because it was included in the core package, if I removed it the tests succeeded). The issue is that one `importlib` loads from the namespace once, it caches the search locations (`importlib.invalidate_caches()` doesn't seem to affect this) so on a subsequent call, if a new path exists, it doesn't get detected. Only the old path is then searched for namespace packages.

File: src/inmanta/loader.py (hunk 12 above)
Comment (sanderr) [reply src/inmanta/loader.py:102]: This is something I went back and forth on. I preferred this because it keeps the `ModuleV2` interface free from legacy methods that will eventually get pruned. But I'm willing to change this if you think otherwise.

Comment (sanderr) [reply src/inmanta/loader.py:102]: Though I now see that Arnaud has already added this to `ModuleV2` where it returns its `setup.cfg` requirements. I'll have to investigate the consequences. The way `SourceInfo.requires` is used, we probably don't want this but I'm not sure yet.

File: src/inmanta/env.py (hunk 6 above)
Comment (wouterdb) [reply src/inmanta/env.py:127]: I would add that as a comment, because it is important.

File: src/inmanta/env.py (hunk 10 above)
Comment (wouterdb) [reply src/inmanta/env.py:211]: also with the additional reset calls that @arnaudsjs added? He did figure out how to reload the namespaces iirc?

File: tests/conftest.py [hunk 15]
```diff
@@ -762,11 +773,14 @@ def setup_func(self, module_dir: Optional[str]):
         self._keep = False
         self.project_dir = tempfile.mkdtemp()
         self.modules_dir = module_dir
+        self.process_env_backup = ProcessEnv.python_path
```
Comment (wouterdb) [inline tests/conftest.py:776]: Why is this here? (i.e. why is it not in the global reset, like the other venv related stuff?)

File: tests/conftest.py [hunk 16]
```diff
@@ -778,11 +792,24 @@ def setup_for_snippet(self, snippet, autostd=True, install_v2_modules: List[Loca
...
         self.setup_for_snippet_external(snippet)
+        loader.unload_inmanta_plugins()
+        loader.PluginModuleFinder.reset()
         project = Project(self.project_dir, autostd=autostd)
         Project.set(project)
         project.use_virtual_env()
+        self._patch_process_env(project)
         self._install_v2_modules(project, install_v2_modules)
 
+    def _patch_process_env(self, project: Project) -> None:
```
Comment (wouterdb) [inline tests/conftest.py:803]: Why is this required (or rather, why is it required now and not in the past)

File: tests/conftest.py [hunk 17]
```diff
@@ -778,11 +792,24 @@ def setup_for_snippet(self, snippet, autostd=True, install_v2_modules: List[Loca
         :param install_v2_modules: Indicates which V2 modules should be installed in the compiler venv
         """
         self.setup_for_snippet_external(snippet)
+        loader.unload_inmanta_plugins()
+        loader.PluginModuleFinder.reset()
```
Comment (wouterdb) [inline tests/conftest.py:796]: is=n't this done by the project.set call just below?

File: tests/conftest.py [hunk 18]
```diff
@@ -778,11 +792,24 @@ def setup_for_snippet(self, snippet, autostd=True, install_v2_modules: List[Loca
         :param install_v2_modules: Indicates which V2 modules should be installed in the compiler venv
         """
         self.setup_for_snippet_external(snippet)
+        loader.unload_inmanta_plugins()
```
Comment (wouterdb) [inline tests/conftest.py:795]: Why do we need this now and not before?

File: tests/conftest.py [hunk 19]
```diff
@@ -1015,3 +1048,103 @@ async def mocked_compiler_service_block(server, monkeypatch):
...
+    """
+    Creates a venv with the latest pip in `${tmpdir}/.venv` where `${tmpdir}` is the directory returned by the
+    `tmpdir` fixture. The venv is activated within the currently running process. This venv is completely decoupled from the
+    active development venv. As a result, any attempts to load new modules from the development venv will fail until cleanup.
+
+    :return: A tuple of the paths to the venv and the Python executable respectively.
+    """
+    venv_dir: py.path.local = tmpdir.join(".venv")
+    python_path: py.path.local = venv_dir.join("bin", "python")
+    venv.create(venv_dir, with_pip=True)
```
Comment (wouterdb) [inline tests/conftest.py:1064]: Who much time do we add to test execution by creating this many venv's?

I wonder if we could do this more efficiently (but this is of secondary importance)

**Review by wouterdb — CHANGES_REQUESTED** [review]
final batch

File: src/inmanta/loader.py (hunk 12 above)
Comment (wouterdb) [reply src/inmanta/loader.py:102]: I like it to be clean now. ModuleV1 will be around for quite a bit into the future and we do have to maintain it.

File: src/inmanta/env.py (hunk 10 above)
Comment (sanderr) [reply src/inmanta/env.py:211]: Yes, I merged in master and some tests started to fail. When I deleted `src/inmanta_plugins/` they succeeded again. Creating the other directory before loading the package the first time (what this method does) solved it.

File: tests/conftest.py (hunk 15 above)
Comment (sanderr) [reply tests/conftest.py:776]: Good point, I should probably move this.

File: src/inmanta/env.py (hunk 6 above)
Comment (sanderr) [reply src/inmanta/env.py:127]: done

File: src/inmanta/env.py (hunk 7 above)
Comment (sanderr) [reply src/inmanta/env.py:138]: done

File: src/inmanta/env.py (hunk 11 above)
Comment (sanderr) [reply src/inmanta/env.py:249]: It looks like it, though not due to this pull request. Should I remove it?

File: src/inmanta/module.py (hunk 14 above)
Comment (sanderr) [reply src/inmanta/module.py]: Done. This required a bit of a refactor of the v1 module source: 0b70c9e.

File: tests/conftest.py (hunk 16 above)
Comment (sanderr) [reply tests/conftest.py:803]: Because with modules v2 we install packages in the wrapping developer environment (not the compiler environment) through `inmanta.env.ProcessEnv`. If we're running tests we don't actually want to install anything in the developer's environment, so we need to mock the wrapping environment.

File: tests/conftest.py (hunk 17 above)
Comment (sanderr) [reply tests/conftest.py:796]: The `reset` isn't but the `unload_inmanta_plugins` is indeed. I'll get rid of that one.

Comment (sanderr) [reply tests/conftest.py:795]: (done)

File: tests/conftest.py (hunk 18 above)
Comment (sanderr) [reply tests/conftest.py]: Removed, see above.

File: tests/conftest.py (hunk 19 above)
Comment (sanderr) [reply tests/conftest.py:1063]: I assume this doesn't affect the total runtime too much since it's only used by a handful of tests. In any case, the `snippetcompiler` fixture already creates a new compiler venv directory for each test that uses it so I believe the performance impact of this will be negligible. I haven't tested it yet though, I'll make sure to do that to be certain.

File: tests/conftest.py (hunk 15 above)
Comment (sanderr) [reply tests/conftest.py]: No reason, it would be better placed in the global reset fixture, I've moved it there.

File: src/inmanta/env.py (hunk 5 above)
Comment (sanderr) [reply src/inmanta/env.py]: Coincidentally I'm using `pkg_resources.working_set` to resolve another review comment, so I've got some idea now.

File: src/inmanta/env.py (hunk 9 above)
Comment (sanderr) [reply src/inmanta/env.py]: I've updated this to use `pkg_resources.working_set` instead.

Comment (sanderr) [reply src/inmanta/env.py]: Nice, this caused a test speedup from +- 30 min to +- 24 min.

File: src/inmanta/env.py (hunk 8 above)
Comment (sanderr) [reply src/inmanta/env.py]: Done

Comment (sanderr) [reply src/inmanta/env.py]: This one did really improve test time (not just because of the Jenkins runner) to the point where this branch's tests now run only a bit slower than those on master, which could be explained by there being more of them.

File: src/inmanta/env.py (hunk 5 above)
Comment (sanderr) [reply src/inmanta/env.py]: (done)

File: src/inmanta/module.py (hunk 13 above)
Comment (sanderr) [reply src/inmanta/module.py]: refactored the whole verify logic to
1. accomodate this
2. verify the modules cache

File: src/inmanta/module.py [hunk 20]
```diff
@@ -276,7 +320,176 @@ def get_file_for_version(self, repo: str, tag: str, file: str) -> str:
...
+        return Requirement.parse(
+            str(module_req).replace(module_req.project_name, cls.get_python_package_name(module_req.project_name))
+        )
+
+    @classmethod
+    def get_namespace_package_name(cls, module_name: str) -> str:
+        return f"{const.PLUGINS_PACKAGE}.{module_name}"
+
+    def install(self, project: Optional["Project"], module_spec: List[InmantaModuleRequirement]) -> Optional["ModuleV2"]:
+        env.process_env.init_namespace(const.PLUGINS_PACKAGE)
```
Comment (wouterdb) [inline src/inmanta/module.py:409]: Why is this required, isn't the namespace package part of this project?

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/module.py (hunk 20 above)
Comment (sanderr) [reply src/inmanta/module.py:409]: I think it can indeed be removed since it's called at `ModuleSourceV2` construction.

Comment (sanderr) [reply src/inmanta/module.py]: done

File: tests/test_env.py [hunk 21]
```diff
@@ -152,3 +162,265 @@ def test_gen_req_file(tmpdir):
         'lorem == 0.1.1, > 0.1 ; python_version < "3.7" and platform_machine == "x86_64" and platform_system == "Linux"'
         in req_lines
     )
+
```
Comment (arnaudsjs) [inline tests/test_env.py:165]: Please add doc string to test cases that don't have one.

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/env.py [hunk 22]
```diff
@@ -288,7 +347,7 @@ def use_virtual_env(self) -> None:
         self._activate_that()
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/env.py:347]: No, the `process_env` singleton is (as the name indicates) meant to represent the environment the compiler process is running in (the root active environment), not the currently activated one. `process_env.install` should never install in the compiler venv.

File: src/inmanta/env.py [hunk 23]
```diff
@@ -47,42 +52,236 @@
...
+
+    def install_from_index(
+        self, requirements: List[Requirement], index_urls: Optional[List[str]] = None, upgrade: bool = False
+    ) -> None:
+        super().install_from_index(requirements, index_urls, upgrade)
+        self.notify_change()
+
+    def install_from_source(self, paths: List[LocalPackagePath]) -> None:
+        super().install_from_source(paths)
+        self.notify_change()
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/env.py:196]: Because `notify_change` is only relevant to notify the active environment. It updates `pkg_resources.working_set`, `sys.site` and `importlib`, which affect the active environment only. I've recently pushed a commit that expanded on the `notify_change` constructor to illustrate its purpose in more detail, have you seen that one? Does it help clarify this?

File: src/inmanta/env.py [hunk 24]
```diff
@@ -47,42 +52,236 @@
...
+            )
+        )
+
+    def get_installed_packages(self, only_editable: bool = False) -> Dict[str, version.Version]:
+        """
+        Return a list of all installed packages in the site-packages of a python interpreter.
 
-    @classmethod
-    def install_from_source(cls, paths: List[LocalPackagePath]) -> None:
+        :param python_interpreter: The python interpreter to get the packages for
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/env.py]: fixed

File: src/inmanta/env.py [hunk 25]
```diff
@@ -47,42 +52,236 @@
...
+
+
 @dataclass
 class LocalPackagePath:
     path: str
     editable: bool = False
 
 
-class ProcessEnv:
+class Environment:
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/env.py]: Good call, renamed.

File: src/inmanta/module.py [hunk 26]
```diff
@@ -276,7 +320,172 @@ def get_file_for_version(self, repo: str, tag: str, file: str) -> str:
...
+            module_names,
+            too_short=ValueError("module_spec should contain at least one requirement"),
+            too_long=ValueError("module_spec should contain requirements for exactly one module"),
+        )
+        return module_name
+
+
+class ModuleV2Source(ModuleSource["ModuleV2"]):
+    def __init__(self, urls: List[str]) -> None:
+        self.urls: List[str] = [url if not os.path.exists(url) else os.path.abspath(url) for url in urls]
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/module.py:388]: If directory exists the index is a local path, in which case I felt it would be more robust to use the abspath rather than a potentially relative path.

File: src/inmanta/module.py [hunk 27]
```diff
@@ -86,6 +88,48 @@
...
+    def __contains__(self, version: str) -> bool:
+        return version in self._requirement
+
+    def __str__(self) -> str:
+        return str(self._requirement).replace("-", "_")
+
+    @classmethod
+    def parse(cls: Type[TInmantaModuleRequirement], spec: str) -> TInmantaModuleRequirement:
+        if "-" in spec:
+            raise ValueError("Invalid inmanta module requirement: inmanta module names use '_', not '-'.")
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/module.py:136]: We can't because `Requirement.parse` transforms all underscores to dashes, so it needs to be checked before constructing the `Requirement` instance, i.e. when parsing. If you think this can be improved I'm open to suggestions.

File: src/inmanta/env.py [hunk 28]
```diff
@@ -47,42 +52,236 @@
...
+            dist_info.key: version.Version(dist_info.version) for dist_info in pkg_resources.working_set
+        }
+        constraint_violations: List[Tuple[Requirement, Optional[version.Version]]] = [
+            (constraint, installed_versions.get(constraint.key, None))
+            for constraint in all_constraints
+            if constraint.key not in installed_versions or str(installed_versions[constraint.key]) not in constraint
+        ]
+
+        for constraint, v in constraint_violations:
+            LOGGER.warning(f"Incompatibility between constraint {constraint} and installed version {v}")
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/env.py]: Applied (manual commit, I additionally converted the f-string to a normal string).

File: src/inmanta/module.py [hunk 29]
```diff
@@ -276,7 +320,176 @@ def get_file_for_version(self, repo: str, tag: str, file: str) -> str:
...
+            raise Exception("PythonRepo instances work with inmanta module names, not Python package names.")
+        package: str = self.get_namespace_package_name(name)
+        mod_spec: Optional[Tuple[Optional[str], Loader]] = env.ActiveEnv.get_module_file(package)
+        if mod_spec is None:
+            return None
+        init, mod_loader = mod_spec
+        if isinstance(mod_loader, loader.PluginModuleLoader):
+            # TODO: test this behavior
+            # module was found in the environment but it is associated with the v1 module loader
+            return None
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/module.py]: I think you've misunderstood the meaning of this condition: it's a perfectly valid state. I've expanded the comment to be more informative. Does that help?

File: src/inmanta/module.py [hunk 30]
```diff
@@ -1597,11 +1901,17 @@ def push(self) -> None:
         print()
 
     @abstractmethod
-    def get_python_requirements_as_list(self) -> List[str]:
+    def get_all_python_requirements_as_list(self) -> List[str]:
         """
-        Return the python requirements specified by the module.
+        Returns all Python requirements specified by the module, including requirements on V2 modules.
         """
-        raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/module.py:1604]: I don't recall doing this, I'm assuming it was an accident (I don't see a good reason to do so). I've restored it.

File: src/inmanta/module.py [hunk 31]
```diff
@@ -919,7 +1137,7 @@ def get_name_from_metadata(cls, metadata: ProjectMetadata) -> str:
         return metadata.name
 
     @property
-    def _install_mode(self) -> InstallMode:
+    def install_mode(self) -> InstallMode:
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/module.py:1145]: With the current annotation we can't really indicate yet which parts of a class are part of the stable API. I've seen the annotation, grepped for `install_mode` in other projects, and looked at the documentation for `inmanta.module.project` (this only mentions `get`, `load`, `set`). So I think it should be safe to update this.

File: tests/test_env.py (hunk 21 above)
Comment (sanderr) [reply tests/test_env.py:165]: Done

File: src/inmanta/env.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/env.py:346]: I've expanded a test case in ce6da334 to verify that the package does indeed get installed in the correct environment.

File: src/inmanta/env.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/env.py:195]: I agree.

**Review by arnaudsjs — APPROVED** [review]
